src/sentry/tasks/deletion/hybrid_cloud.py

```diff
@@ -0,0 +1,214 @@
+"""
+Executes the eventually consistent cascades dictated by HybridCloudForeignKey fields on application models.
+
+One job schedules on a regular interval the execute of smaller, per HCFK column jobs that each do a small chunk of work,
+and then possibly reschedule to keep processing as necessary to stay caught up.
+
+Notably, this job is only responsible for cascading to models that are related to deletions that have occurred in the
+opposing silo and are stored in Tombstone rows.  Deletions that are not successfully synchronized via Outbox to a
+Tombstone row will not, therefore, cascade to any related cross silo rows.
+"""
+from hashlib import sha1
+from typing import Any, List, Tuple, Type
+from uuid import uuid4
+
+import sentry_sdk
+from django.apps import apps
+from django.db.models import Manager, Max
+
+from sentry.db.models.fields.hybrid_cloud_foreign_key import HybridCloudForeignKey
+from sentry.models import TombstoneBase
+from sentry.silo import SiloMode
+from sentry.tasks.base import instrumented_task
+from sentry.utils import json, metrics, redis
+
+
+def deletion_silo_modes() -> List[SiloMode]:
+    cur = SiloMode.get_current_mode()
+    result: List[SiloMode] = []
+    if cur != SiloMode.REGION:
+        result.append(SiloMode.CONTROL)
+    if cur != SiloMode.CONTROL:
+        result.append(SiloMode.REGION)
+    return result
+
+
+def get_watermark_key(prefix: str, field: HybridCloudForeignKey) -> str:
+    return f"{prefix}.{field.model._meta.db_table}.{field.name}"
+
+
+def get_watermark(prefix: str, field: HybridCloudForeignKey) -> Tuple[int, str]:
+    with redis.clusters.get("default").get_local_client_for_key("deletions.watermark") as client:
+        key = get_watermark_key(prefix, field)
+        v = client.get(key)
+        if v is None:
+            result = (0, uuid4().hex)
+            client.set(key, json.dumps(result))
+            return result
+        return tuple(json.loads(v))
+
+
+def set_watermark(
+    prefix: str, field: HybridCloudForeignKey, value: int, prev_transaction_id: str
+) -> None:
+    with redis.clusters.get("default").get_local_client_for_key("deletions.watermark") as client:
+        client.set(
+            get_watermark_key(prefix, field),
+            json.dumps((value, sha1(prev_transaction_id.encode("utf8")).hexdigest())),
+        )
+    metrics.gauge(
+        "deletion.hybrid_cloud.low_bound",
+        value,
+        tags=dict(
+            table_name=field.model._meta.db_table,
+            field_name=field.name,
+            watermark=prefix,
+        ),
+    )
```

Comment:
If/When we lose the watermarks for tombstones (because of a redis failure/blip) is the recovery path that we reprocess all the existing tombstones again?

Reply:
Yes, but it will be significantly faster; the queries will ultimately yield no positives and no deletion will trigger.
Essentially "reprocess" will be to query every HCFK against the tombstones once, in batches.